src/ControllerPose.cpp

```diff
@@ -1,152 +1,92 @@
+    // multiply the rotation matrix by the offset vector set that is the offset of the controller
+    // relative to the hand
+    vr::HmdVector3_t vectorOffset =
+        MultiplyMatrix(controllerRotationMatrix, m_poseConfiguration.offsetVector);
 
-            newPose.vecVelocity[0] = trackedDevicePoses[m_shadowControllerId].vVelocity.v[0];
-            newPose.vecVelocity[1] = trackedDevicePoses[m_shadowControllerId].vVelocity.v[1];
-            newPose.vecVelocity[2] = trackedDevicePoses[m_shadowControllerId].vVelocity.v[2];
+    // combine these positions to get the resultant position
+    vr::HmdVector3_t newControllerPosition = CombinePosition(controllerMatrix, vectorOffset);
 
-            newPose.poseIsValid = true;
-            newPose.deviceIsConnected = true;
+    newPose.vecPosition[0] = newControllerPosition.v[0];
+    newPose.vecPosition[1] = newControllerPosition.v[1];
+    newPose.vecPosition[2] = newControllerPosition.v[2];
 
-            newPose.result = vr::TrackingResult_Running_OK;
+    // Multiply rotation quaternions together, as the controller may be rotated relative to the
+    // hand
+    newPose.qRotation = MultiplyQuaternion(GetRotation(controllerMatrix),
+                                           m_poseConfiguration.angleOffsetQuaternion);
 
-            newPose.poseTimeOffset = m_poseConfiguration.poseOffset;
-        } else {
-            newPose.poseIsValid = false;
-            newPose.deviceIsConnected = true;
-            newPose.result = vr::TrackingResult_Uninitialized;
-        }
+    // Copy other values from the controller that we want for this device
+    newPose.vecAngularVelocity[0] = trackedDevicePoses[m_shadowControllerId].vAngularVelocity.v[0];
+    newPose.vecAngularVelocity[1] = trackedDevicePoses[m_shadowControllerId].vAngularVelocity.v[1];
+    newPose.vecAngularVelocity[2] = trackedDevicePoses[m_shadowControllerId].vAngularVelocity.v[2];
 
-    } else {
-        newPose.result = vr::TrackingResult_Uninitialized;
-        newPose.deviceIsConnected = false;
-        DiscoverController();
-    }
+    newPose.vecVelocity[0] = trackedDevicePoses[m_shadowControllerId].vVelocity.v[0];
+    newPose.vecVelocity[1] = trackedDevicePoses[m_shadowControllerId].vVelocity.v[1];
+    newPose.vecVelocity[2] = trackedDevicePoses[m_shadowControllerId].vVelocity.v[2];
 
-    return newPose;
-}
+    newPose.poseIsValid = true;
+    newPose.deviceIsConnected = true;
 
-//Checks if a role is the opposite hand from the current hand
-bool ControllerPose::IsOtherRole(int32_t test) {
-    if (m_shadowDeviceOfRole == vr::ETrackedControllerRole::TrackedControllerRole_RightHand)
-        return test == vr::ETrackedControllerRole::TrackedControllerRole_LeftHand;
-    else
-        return test == vr::ETrackedControllerRole::TrackedControllerRole_RightHand;
-}
+    newPose.result = vr::TrackingResult_Running_OK;
 
-void ControllerPose::DiscoverController() {
```

Comment:
Would it be a good idea to keep the old DiscoverController method inside as a deprecated way of pose tracking for debug purposes, etc?

Reply:
eh, what we had before was a bit of a yucky solution - if we wanted to we could do the same sort of priority queue in the overlay though... might have to experiment with it though.

Reply:
Ah, yeah fair enough